src/groups/special_euclidean.jl

```diff
@@ -119,20 +108,42 @@ function affine_matrix(G::SpecialEuclidean{n}, p) where {n}
 end
 affine_matrix(::SpecialEuclidean{n}, p::AbstractMatrix) where {n} = p
 @generated function affine_matrix(::GT, ::Identity{GT}) where {n,GT<:SpecialEuclidean{n}}
-    return SDiagonal{n}(I)
+    s = maybesize(Size(n, n))
+    s isa Size && return SDiagonal{n,Float64}(I)
```

Comment:
Does it have to be a generated function? It's likely that constant propagation, inlining and optimization will make this fast anyway. And if not, `Diagonal{Float64}(I, n)` should probably be constructed at runtime (quoted return `:(Diagonal{Float64}(I, n))`) instead of compile time in case someone wants to modify it.

Reply:
It doesn't need to be generated. For SE(3), there's no speed-up. It's just for higher dimensional SE(N) that you see a speed-up, but those are unlikely to be used.

Can you explain the benefit of returning quoted? How would one modify it?

Reply:
Here is an example:
```julia
julia> @generated f(x) = Diagonal{Float64}(I, 3)
f (generic function with 1 method)

julia> using LinearAlgebra

julia> A = f(10)
3×3 Diagonal{Float64,Array{Float64,1}}:
 1.0   ⋅    ⋅ 
  ⋅   1.0   ⋅ 
  ⋅    ⋅   1.0

julia> A[1] = 10
10

julia> f(20)
3×3 Diagonal{Float64,Array{Float64,1}}:
 10.0   ⋅    ⋅ 
   ⋅   1.0   ⋅ 
   ⋅    ⋅   1.0

julia> @generated g(x) = :(Diagonal{Float64}(I, 3))
g (generic function with 1 method)

julia> B = g(10)
3×3 Diagonal{Float64,Array{Float64,1}}:
 1.0   ⋅    ⋅ 
  ⋅   1.0   ⋅ 
  ⋅    ⋅   1.0

julia> B[1] = 20
20

julia> g(99)
3×3 Diagonal{Float64,Array{Float64,1}}:
 1.0   ⋅    ⋅ 
  ⋅   1.0   ⋅ 
  ⋅    ⋅   1.0
```
The returned quoted expression is evaluated when the generated function is called, so it makes a difference when a mutable object is allocated.

Reply:
Whoah, that is really good to know! You probably just saved me hours of future work hunting down bugs that would be caused by this. Thanks!

Reply:
No problem :slightly_smiling_face: . Generated functions are quite tricky to write, so I prefer to keep them as simple as possible.